### PR TITLE
Add instructions for external computation

### PR DESCRIPTION
**Pull request by JuanCoRo**
This PR adds a new Foundry project under `docs/state-diff` containing
- An explainer (in the README) on how to offload computation to Forge and then introduce it to the proofs
- A running example of how this is done, to be found under `docs/state-diff/test/kontrol`

### CODE REVIEW

File: docs/state-diff/README.md [hunk 1]
```diff
@@ -0,0 +1,185 @@
+ 🤖⚡ External Computation with Kontrol ⚡🤖
+=================================
+**Injecting Foundry Execution into Proofs**
```
Comment (palinatolmach) [inline docs/state-diff/README.md:3]:
```suggestion
**Injecting Foundry Execution into Proofs for Improved Performance**
```
That's a draft suggestion, but I think it might be helpful to incorporate some information about why would we want to do it here, e.g. say that it's helpful for projects with complex but concrete setup process.

File: docs/state-diff/README.md [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+This folder is an example for including external computation with Foundry into Kontrol proofs.
```
Comment (palinatolmach) [inline docs/state-diff/README.md:5]: Maybe "including results of Foundry execution into the Kontrol initial state"?
I would just phrase it in a bit more details manner. It becomes clear what "external computation with Foundry" is in the next paragraph, but we can probably have a more clear term here too.

File: docs/state-diff/README.md [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+To achieve this, there are several files involved and the process has several steps. If you encounter any problems while following what is described here, please reach out in [Discord](https://discord.gg/CurfmXNtbN), we'll be happy to assist!
```
Comment (palinatolmach) [inline docs/state-diff/README.md:6]:
```suggestion
To achieve this, there are several files involved and the process has several steps. If you encounter any problems while following these instructions, please reach out in [Discord](https://discord.gg/CurfmXNtbN), we'll be happy to assist!
```

Comment (palinatolmach) [reply docs/state-diff/README.md:6]: Also, maybe I'd rephrase the first sentence to say something like "This projects contains the files and description of steps needed for this process". :)

File: docs/state-diff/README.md [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+To achieve this separation we'll introduce a new Foundry profile, usually called `kprove`. We also need a fresh folder where the proofs will live. If, for instance, that folder is `test/kontrol/proofs`, the resulting profile would look as follows:
+```toml
+[profile.kprove]
+src = 'test/kontrol/proofs'
+out = 'kout-proofs'
+test = 'test/kontrol/proofs'
+script = 'test/kontrol/proofs'
+```
+
+### ⚠️  Special `out` directory ⚠️
```
Comment (palinatolmach) [inline docs/state-diff/README.md:32]:
```suggestion
### ⚠️ Special `out` directory ⚠️
```

File: docs/state-diff/README.md [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+Note that we're also setting a different `out` directory named `kout-proofs`. Because of this, any Kontrol related command will have to be executed in the context of the `kprove` profile. This can be achieved either by `export FOUNDRY_PROFILE=kprove` or by prepending any Kontrol command with that flag (e.g., `FOUNDRY_PROFILE=kprove kontrol list`).
```
Comment (palinatolmach) [inline docs/state-diff/README.md:34]:
```suggestion
Note that we're also setting a different `out` directory named `kout-proofs`. Because of this, any Kontrol-related command will have to be executed in the context of the `kprove` profile. This can be achieved either by `export FOUNDRY_PROFILE=kprove` or by prepending any Kontrol command with that flag (e.g., `FOUNDRY_PROFILE=kprove kontrol list`).
```

File: docs/state-diff/README.md [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+
+## 🖋️ Record your execution 🖋️
+
+To record your execution and save its output to a JSON file, simply use the modifier `recordStateDiff` in the [`test/kontrol/state-diff/record-state-diff/RecordStateDiff.sol`](./test/kontrol/state-diff/record-state-diff/RecordStateDiff.sol) file. That is, the initial set up of your proofs has to be run in a function with the `recordStateDiff` modifier. An example of this can be found in [`test/kontrol/state-diff/proof-initialization.sol`](test/kontrol/state-diff/proof-initialization.sol).
+
+On top of that, the `save_address` function allows to save the name of the deployed contract and their addresses into a separate JSON file. This will be crucial to easily write the symbolic property tests later on.
+
+### 🏷️ Name the files 🏷️
+
+There are two different JSON files that can be created, one containing the recorded state updates and another one with the saved names of important addresses. On top of that, we have to define in which directory do these files live. For simplicity, we've made it so that these parameters are set with environment variables:
```
Comment (palinatolmach) [inline docs/state-diff/README.md:44]: Maybe I'd clarify what the "important addresses" are — the ones that are meant to be used in the tests?

File: docs/state-diff/README.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+- `STATE_DIFF_NAME`: Name of the JSON containing the state updates. Example: `StateDiff.json`
+- `ADDR_NAMES`: Name of the JSON containing the saved names of relevant addresses. Example: `AddressNames.json`
+- `STATE_DIFF_DIR`: Path relative to the foundry root dir where the files will be stored. Example: `state-diff`
```
Comment (palinatolmach) [inline docs/state-diff/README.md:47]: Just for consistency: 
```suggestion
- `STATE_DIFF_DIR`: Path relative to the Foundry root dir where the files will be stored. Example: `state-diff`
```

File: docs/state-diff/README.md [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+
+Run your function containing the initial set up of your proofs (`counterBed` or `counterBedNamed` in our [example](./test/kontrol/state-diff/proof-initialization.sol)) with `forge script state-diff/proof-initialization.sol:CounterBed --sig counterBed --ffi`. Running it with `forge test` will also work, but only if its name starts with `test`. Notice the `--ffi` flag: we use `mkdir` and `touch` to handle the cases where the state diff files don't yet exist.
+
+### 🧼 Clean the State Diff file 🧼
+
+Currently, the produced state diff JSON is escaped when generated. Run the [`clean_json.py`](test/kontrol/scripts/json/clean_json.py) script on the generated state diff file to unescape it. Example: `python3 test/kontrol/scripts/json/clean_json.py state-diff/StateDiff.json`
+
+### 📜 Scripting all of this 📜
+
+Phew! This is nothing short of a mouthful of steps! In case you don't want to generate the necessary scripts by manally repeating all that we have gone through, we provide the [`record-state-diff.sh`](test/kontrol/scripts/record-state-diff.sh) script automatically generate the JSON files.
```
Comment (palinatolmach) [inline docs/state-diff/README.md:69]: I'd just say
```suggestion
Phew! This is nothing short of a mouthful of steps! In case you don't want to manually repeat all that we have gone through, we provide the [`record-state-diff.sh`](test/kontrol/scripts/record-state-diff.sh) script automatically generate the JSON files.
```
Please feel free to ignore the suggestion! But please note the typo in "manually" :)

File: docs/state-diff/README.md [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        }
+    }
+
+function test_multiple_counters() public {
+        for(uint256 i; i <= 9; ++i){
+            require(counters[i].number() == i, "Bad number");
+        }
+    }
+```
+we can offload the computation of the `setUp` function (of any function, really) to Foundry, and then provide to Kontrol the file containing what state updates occurred during the execution of `setUp`. Let's see how this transformation is done.
```
Comment (palinatolmach) [inline docs/state-diff/README.md:91]:
```suggestion
we can offload the computation of the `setUp` function (of any function, really) to Foundry, and then provide Kontrol with the file containing the state updates that occurred during the execution of `setUp`. Let's see how this transformation is done.
```

File: docs/state-diff/README.md [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+    for (uint256 i; i <= 9; ++i) {
+        counter = new Counter();
+        counter.setNumber(i);
+        save_address(address(counter), string.concat("counter", vm.toString(i)));
+    }
+}
+```
+Notice that we're saving the addresses via `save_address`, not via the `Counter[]`. 
+
+The next step is to use Kontrol's `load-state-diff` command. An example on how to use it can be found in [the provided script](test/kontrol/scripts/record-state-diff.sh). For our running example it can be called as follows
```
Comment (palinatolmach) [inline docs/state-diff/README.md:105]: For consistency:
```suggestion
The next step is to use Kontrol's `load-state-diff` command. An example on how to use it can be found in [the provided script](test/kontrol/scripts/record-state-diff.sh). For our running example it can be called as follows:
```

File: docs/state-diff/README.md [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+address internal constant counter1Address = 0x2e234DAe75C793f67A35089C9d99245E1C58470b;
+address internal constant counter0Address = 0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f;
+address internal constant counter3Address = 0x5991A2dF15A8F6A256D3Ec51E99254Cd3fb576A9;
+address internal constant counter7Address = 0xA4AD4f68d0b91CFD19687c881e50f3A00242828c;
+address internal constant counter9Address = 0xD6BbDE9174b1CdAa358d2Cf4D57D1a9F7178FBfF;
+address internal constant counter2Address = 0xF62849F9A0B5Bf2913b396098F7c7019b51A820a;
+address internal constant counter5Address = 0xa0Cb889707d426A7A386870A03bc70d1b0697598;
+address internal constant counter4Address = 0xc7183455a4C133Ae270771860664b6B7ec320bB1;
+```
+This means that for [our test](test/kontrol/proofs/Counter.k.sol) we'll inherit `InitialState.sol` and add the necessary interface to each address:
```
Comment (palinatolmach) [inline docs/state-diff/README.md:128]:
```suggestion
This means that we'll make [our test](test/kontrol/proofs/Counter.k.sol) inherit `InitialState.sol` and add the necessary interface to each address:
```

File: docs/state-diff/README.md [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+    counters.push(Counter(address(counter3Address)));
+    counters.push(Counter(address(counter4Address)));
+    counters.push(Counter(address(counter5Address)));
+    counters.push(Counter(address(counter6Address)));
+    counters.push(Counter(address(counter7Address)));
+    counters.push(Counter(address(counter8Address)));
+    counters.push(Counter(address(counter9Address)));
+}
+```
+Note that in the previous `setUp` we were deploying contracts + calling `setNumber` on each contract and storing them in the `Counters[]` array. Here we're only casting the addresses with the right interface and storing them to `counters`.
```
Comment (palinatolmach) [inline docs/state-diff/README.md:145]:
```suggestion
Note that in the previous `setUp` we were deploying contracts + calling `setNumber` on each contract and storing them in the `Counters[]` array. Here we're only casting the addresses with the right interface and storing them in `counters`.
```

File: docs/state-diff/README.md [hunk 13]
```diff
@@ -0,0 +1,185 @@
...
+
+**A note on interfaces**: One of the goals of this technique is to isolate the verified bytecode from the sourcecode producing it. Therefore `Counter` here is just an interface of the actual `Counter` contract. The interface can be found [here](test/kontrol/proofs/utils/Interfaces.sol). To see more clearly how this is brought together, see the test [here](test/kontrol/proofs/Counter.k.sol).
```
Comment (palinatolmach) [inline docs/state-diff/README.md:147]:
```suggestion
**A note on interfaces**: One of the goals of this technique is to isolate the verified bytecode from the source code producing it. Therefore `Counter` here is just an interface of the actual `Counter` contract. The interface can be found [here](test/kontrol/proofs/utils/Interfaces.sol). To see more clearly how this is brought together, see the test [here](test/kontrol/proofs/Counter.k.sol).
```

File: docs/state-diff/README.md [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    counters.push(Counter(address(counter3Address)));
+    counters.push(Counter(address(counter4Address)));
+    counters.push(Counter(address(counter5Address)));
+    counters.push(Counter(address(counter6Address)));
+    counters.push(Counter(address(counter7Address)));
+    counters.push(Counter(address(counter8Address)));
+    counters.push(Counter(address(counter9Address)));
+}
+```
+And the following tests that asserts some property for each of the stored addresses:
```
Comment (palinatolmach) [inline docs/state-diff/README.md:172]:
```suggestion
And the following test that asserts some property of each of the stored addresses:
```

File: docs/state-diff/README.md [hunk 15]
```diff
@@ -0,0 +1,185 @@
...
+            require(counters[i].number() == i, "Bad number");
+        }
+    }
+```
+All that is left is to tell Kontrol which state updates need to be loaded before executing `prove_multiple_counters`. This is done via the `--init-node-from` flag of the `kontrol prove` command. The `--init-node-from` flag expects a JSON file containing the state updates and will load them to the initial state of the proof. This will make Kontrol aware of all the changes that occurred during the execution of the original `setUp` function, with none of the computation.
+
+Hence, to successfully execute the above function in Kontrol we'll have to execute:
+1. `FOUNDRY_PROFILE=kprove kontrol build`
+2. `FOUNDRY_PROFILE=kprove kontrol prove --match-test prove_multiple_counters --init-node-from state-diff/StateDiff.json`
+Note that running this in the context of the `kprove` profile is crucial, since it points to the isolated folder that will contain all the necessary bytecode.
```
Comment (palinatolmach) [inline docs/state-diff/README.md:185]: Maybe you can show the result? Like the `PROOF PASSED: ...` log? I think we can use some closure in this tutorial :)

File: docs/state-diff/README.md (hunk 1 above)
Comment (JuanCoRo) [reply docs/state-diff/README.md:3]: I added reference to the usability in the subtitle, but since the readme is long enough, I thought to defer the motivation of why would anyone endure these steps to the blog post. Does that sound alright?

File: docs/state-diff/README.md (hunk 2 above)
Comment (JuanCoRo) [reply docs/state-diff/README.md:5]: I like this!

File: docs/state-diff/README.md (hunk 3 above)
Comment (JuanCoRo) [reply docs/state-diff/README.md:6]: Liking this as well :)

File: docs/state-diff/README.md (hunk 6 above)
Comment (JuanCoRo) [reply docs/state-diff/README.md:44]: Good catch. Gotta explain what these VIP addresses are!

File: docs/state-diff/README.md (hunk 8 above)
Comment (JuanCoRo) [reply docs/state-diff/README.md:69]: Oh my!  It will not come the day I don't do typos :)

File: docs/state-diff/README.md (hunk 2 above)
Comment (JuanCoRo) [reply docs/state-diff/README.md]: I reworked the entire intro! Makes sense to get the point across as to why would we ever want to do all of that!

File: docs/state-diff/README.md (hunk 1 above)
Comment (palinatolmach) [reply docs/state-diff/README.md]: Yes, agreed!

File: docs/state-diff/README.md (hunk 2 above)
Comment (palinatolmach) [reply docs/state-diff/README.md]: Thank Juan, I think it looks great!

**Review by anvacaru — APPROVED** [review]
LGTM! I'd say to wait for @palinatolmach's approve as well.

**Review by palinatolmach — APPROVED** [review]
Looks good to me too!
